src/block_service/hdfs/hdfs_service.cpp

```diff
@@ -0,0 +1,454 @@
+                                     const ls_callback &cb,
+                                     dsn::task_tracker *tracker = nullptr)
+{
+    ls_future_ptr tsk(new ls_future(code, cb, 0));
+    tsk->set_tracker(tracker);
+
+    auto list_dir_background = [this, req, tsk]() {
+        std::string path = dsn::utils::filesystem::path_combine(_hdfs_path, req.dir_name);
+        ls_response resp;
+
+        if (hdfsExists(_fs, path.c_str()) == -1) {
+            derror_f("HDFS list directory failed: path {} not found.", path);
+            resp.err = ERR_OBJECT_NOT_FOUND;
+            tsk->enqueue_with(resp);
+            return;
+        }
+
+        hdfsFileInfo *dir_info = hdfsGetPathInfo(_fs, path.c_str());
+        if (dir_info == nullptr) {
+            derror_f("HDFS get path {} failed.", path);
+            resp.err = ERR_FS_INTERNAL;
+        } else if (dir_info->mKind == kObjectKindFile) {
+            derror_f("HDFS list directory failed, {} is not a directory", path);
+            resp.err = ERR_INVALID_PARAMETERS;
+        } else {
+            int entries = 0;
+            hdfsFileInfo *info = hdfsListDirectory(_fs, path.c_str(), &entries);
+            if (info == nullptr) {
+                derror_f("HDFS list directory {} failed.", path);
+                resp.err = ERR_FS_INTERNAL;
+            } else {
+                for (int i = 0; i < entries; i++) {
+                    ls_entry tentry;
+                    tentry.entry_name = get_entry_name(std::string(info[i].mName));
+                    tentry.is_directory = (info[i].mKind == kObjectKindDirectory);
+                    resp.entries->emplace_back(tentry);
+                }
+                resp.err = ERR_OK;
+            }
+            hdfsFreeFileInfo(info, entries);
+        }
+        hdfsFreeFileInfo(dir_info, 1);
+        tsk->enqueue_with(resp);
+    };
+
+    dsn::tasking::enqueue(LPC_HDFS_SERVICE_CALL, nullptr, list_dir_background);
+    return tsk;
+}
+
+dsn::task_ptr hdfs_service::create_file(const create_file_request &req,
+                                        dsn::task_code code,
+                                        const create_file_callback &cb,
+                                        dsn::task_tracker *tracker = nullptr)
+{
+    create_file_future_ptr tsk(new create_file_future(code, cb, 0));
+    tsk->set_tracker(tracker);
+    std::string hdfs_file = dsn::utils::filesystem::path_combine(_hdfs_path, req.file_name);
+
+    if (req.ignore_metadata) {
+        create_file_response resp;
+        resp.err = ERR_OK;
+        resp.file_handle = new hdfs_file_object(this, hdfs_file);
+        tsk->enqueue_with(resp);
+        return tsk;
+    }
+
+    auto create_file_in_background = [this, req, hdfs_file, tsk]() {
+        create_file_response resp;
+        dsn::ref_ptr<hdfs_file_object> f = new hdfs_file_object(this, hdfs_file);
+        resp.err = f->get_file_meta();
+        if (resp.err == ERR_OK || resp.err == ERR_OBJECT_NOT_FOUND) {
```

Comment:
Why creating with `ERR_OBJECT_NOT_FOUND` is success?

Reply:
We call this function just to create a `hdfs_file_object`, we may create a new file to write or read from an existing file in the following steps.

Reply:
```suggestion
        // Just to create a hdfs_file_object locally. The file may not appear on HDFS
        // immediately after this call.
        if (resp.err == ERR_OK || resp.err == ERR_OBJECT_NOT_FOUND) {
```